{% extends "dbspecific.html" %}

{% block title %} | {{ protein.gene_name }}{% endblock %}

{% block header %}
{{ super() }}
<link href="{{ prefix }}/static/css/bootstrap-table.min.css" rel="stylesheet">
<link href="{{ prefix }}/static/css/bootstrap-table-export.min.css" rel="stylesheet">
<link href="{{ prefix }}/static/css/protein.css" rel="stylesheet">
<script src="{{ prefix }}/static/js/protein.js"></script>
<script src="{{ prefix }}/static/js/bootstrap-table.min.js"></script>
<script src="{{ prefix }}/static/js/orthotable.js"></script>
<script>
	const access="{{protein.access}}";
	const dbSize="{{protein.taxon|length}}";
	const seqLength={{protein.length}};
	const model=Boolean("{{model}}");
</script>
{% endblock %}

{% block content %}
<h1>{{ protein.uniprot_id }} <span class="badge" title="Database version {{ db.version }}">{{ db.version }}</span></h1>
<p>
	<b>Gene name</b> {{ protein.gene_name }}
</p>
<p>
	<b>Accession number</b> {{ protein.access }}
</p>
<p>
	<b>Description</b> {{ protein.short_desc }}
</p>
<p>
	<b>Links</b> <a target="_blank" href="http://uniprot.org/uniprot/{{ protein.access }}">Uniprot</a>
</p>
<p>
	<b>Organism</b> <a target="_blank" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={{ protein.taxid }}">{{ protein.species }}</a>
</p>
<p>
	<b>Length</b> {{ protein.length }}
</p>
<div class="panel-group">
	<div class="panel panel-default">
		<div class="panel-heading">
			<h4 class="panel-title">
				<a data-toggle="collapse" href="#collapseSeq">Sequence</a>
			</h4>
		</div>
		<div id="collapseSeq" class="panel-collapse collapse">
			<div class="panel-body">
				<div id='sequence' style="font-family: monospace; display:inline-block; width:60ch; word-wrap: break-word; vertical-align: top; margin-left: 1em">
					&gt;{{protein.description}}
					<br />{{protein.sequence}}
				</div>
			</div>
		</div>
	</div>
	<div class="panel panel-default">
		<div class="panel-heading">
			<h4 class="panel-title">
				<a id="show-annot" data-toggle="collapse" href="#collapseGO">Gene Ontology</a> <a data-toggle='modal' data-target='#GoHelpWindow'> <span class='glyphicon glyphicon-info-sign'></span></a>
			</h4>
		</div>
		<div id="collapseGO" class="panel-collapse collapse">
			<div class='panel-body'>
				Loading...
			</div>
		</div>
	</div>
	<div class="panel panel-default">
		<div class="panel-heading">
			<h4 class="panel-title">
				<a id="show-dom" data-toggle="collapse" href="#collapseDoms">Domains</a> <a data-toggle='modal' data-target='#InterproHelpWindow'><span class='glyphicon glyphicon-info-sign'></span></a>
			</h4>
		</div>
		<div id="collapseDoms" class="panel-collapse collapse">
			<div class="panel-body">
				<div id="interpro" style="font-family: monospace; display:inline-block; width:60ch; word-wrap: break-word; vertical-align: top; margin-left: 1em">Loading...</div>
			</div>
		</div>
	</div>
	<hr />
	<div>
		{% if protein.has_transverse and db.name != 'Transverse' %}
		<form action="{{prefix}}/Transverse/protein/{{protein.access}}">
			<button class="btn btn-success"  type='submit'>View three domains</button>
		</form>
		{% elif database == 'Transverse' %}
		<form action="{{prefix}}/{{database}}/protein/{{protein.access}}">
			<button class="btn btn-success"  type='submit'>View {{database}} only</button>
		</form>
		{% endif %}
	</div>
	<div id='GoHelpWindow' class='modal fade' role='dialog'>
		<div class='modal-dialog'>
			<div class='modal-content'>
				<div class='modal-header'>
					<button type='button' class='close' data-dismiss='modal'>&times;</button>
					<h4 class='modal-title'>Help</h4>
				</div>
				<div class='modal-body'>
					<p><b>What is the source the Gene Ontology annotation?</b></p>
					<p>The Gene Ontology annotations are retrieved from the EBI's QuickGO webservice. Further informations are available in the <a href=" {{prefix}}/external_Data">external data sources section</a> </p>
					<p><b>What is the quality of these annotations?</b></p>
					<p>For now, all GO term annotations, regardless of their evidence code, are displayed. No quality filter are used to provide the widest possible range of annotations. You can check the labels near each annotation for evidence details: manually reviewed annotations are displayed in blue, while the automatic ones are in grey.</p>  
				</div>

				<div class='modal-footer'>
					<button type='button' class='btn btn-default' data-dismiss='modal'>Close</button>
				</div>
			</div>
		</div>
	</div>
	<div id='InterproHelpWindow' class='modal fade' role='dialog'>
		<div class='modal-dialog'>
			<div class='modal-content'>
				<div class='modal-header'>
					<button type='button' class='close' data-dismiss='modal'>&times;</button>
					<h4 class='modal-title'>Help</h4>
				</div>
				<div class='modal-body'>
					<p><b>What is the source the domain annotation?</b></p>
					<p>The domain annotations are dynamically queried from the Interpro database, using EBI webservice.</p>
					<p>The source of these annotations is represented by a color code (see legend). You can access the corresponding Uniprot entries in Interpro by clicking on the synthetic representation of each domain.</p>  
				</div>

				<div class='modal-footer'>
					<button type='button' class='btn btn-default' data-dismiss='modal'>Close</button>
				</div>
			</div>
		</div>
	</div>
	<div id='distanceHelpWindow' class='modal fade' role='dialog'>
		<div class='modal-dialog'>
			<div class='modal-content'>
				<div class='modal-header'>
					<button type='button' class='close' data-dismiss='modal'>&times;</button>
					<h4 class='modal-title'>Help</h4>
				</div>
				<div class='modal-body'>
					<p><b>What are the distances?</b></p>
					<p>The distances shown here are Jaccard distances, which are adapted to binary profiles (presence/absence).</p>
					<p><b>How are they computed?</b></p>
					<p>Distances between two protein of a given species are computed on phylogenetic profile of each protein in model species of the considered database.</p>
					<p><b>Threshold</b></p>
					<p>The goal of this section is to provide proteins with similar distributions to the query, to that end, we set a maximum threshold at 0.4</p>
				</div>
				<div class='modal-footer'>
					<button type='button' class='btn btn-default' data-dismiss='modal'>Close</button>
				</div>
			</div>
		</div>
	</div>
	<div class="panel-group">
		<div class="panel panel-default">
			<div class="panel-heading">
				<h1 class="panel-title">
					<a data-toggle="collapse" href="#collapseOrthologs"><b>Orthologs and taxonomic distribution</b></a>
				</h1>
			</div>
			<div id="collapseOrthologs" class="panel-collapse collapse in">
				<div class="panel-body"> 
                    {% if db.has_models %}
					<ul class="nav nav-tabs">
						<li {% if model %}class="active" {% endif %}> <a href="{{prefix}}/{{db.name}}/protein/{{db.version}}/{{protein.access}}">Model Organisms</a></li>
						<li {% if not model %}class="active" {% endif %}><a href="{{prefix}}/{{db.name}}/protein/{{db.version}}/{{protein.access}}/full">Whole domain</a></li>
					</ul>        
                    {% endif %}
					<div class="tax_summary" id='band'>
					</div>
					<table id="orthotable" class="table" data-toolbar="#toolbar" data-toggle ="table" data-url="{{prefix}}/{{db.name}}/orthologs/{{protein.access}}{% if not model %}/full{% endif %}" data-sort-name="taxoDist" data-sort-order="asc" data-show-toggle="true" data-show-export="true" data-show-columns="true" data-search="true" data-search-placeholder="Search protein or taxon" data-click-to-select="true" data-select-item-name="input" data-id-field="id">
						<div id="toolbar">
							<button id="align-button" class="btn btn-primary" onclick="launchAlignment()">Align with PipeAlign</button>
							<button id="fasta" class='btn btn-primary'>Get FASTA</button>
							{#<button id="orthoxml" class='btn btn-primary'>Get OrthoXML</button>#}
						</div>
						<thead>
							<tr>
								<th data-field="Selected" data-checkbox="true"></th>
								<th data-field="type" data-sortable="true">Relationship</th>
								<th data-field="inparalogs" data-switchable="false" data-formatter="sequence_formatter">Query &amp; inparalogs</th>
								<th data-field="orthologs" data-switchable="false" data-formatter="sequence_formatter">Orthologs</th>
								<th data-visible="false" data-formatter="size_formatter" data-field="size">Orthologs length</th>
								<th data-field="species" data-sortable="true" data-formatter="species_formatter">Species</th>
								<th data-visible="false" data-switchable="false" data-field="species" search-formatter="taxid_search_formatter">Taxid</th>
								<th data-visible="false" data-switchable="false" data-field="taxoDist" data-sortable="true"/>
								<th data-field="reducedTaxonomy" data-formatter="taxonomy_formatter">Taxonomy</th>
								<th data-visible="false" data-switchable="false" data-field="fullTaxonomy" data-formatter="taxonomy_formatter">Whole Taxonomy</th>
								<th data-visible="false" data-switchable="false" data-field="taxoDist" />
							</tr>
						</thead>
					</table>
				</div>
			</div>
		</div>
	</div>
</div>
{% endblock %}
